manta-pay/src/crypto/poseidon/mod.rs

```diff
@@ -0,0 +1,447 @@
+            let linear_combination = state
+                .iter()
+                .enumerate()
+                .map(|(j, elem)| S::mul_const(elem, &self.mds_matrix[S::WIDTH * i + j], compiler))
+                .collect::<Vec<_>>();
+            next.push(
+                linear_combination
+                    .into_iter()
+                    .reduce(|acc, next| S::add(&acc, &next, compiler))
+                    .unwrap(),
+            );
+        }
+        mem::swap(&mut next.into_boxed_slice(), &mut state.0);
+    }
+
+    /// Computes a full round at the given `round` index on the internal permutation `state`.
+    #[inline]
+    pub fn full_round(&self, round: usize, state: &mut State<S, COM>, compiler: &mut COM) {
+        let keys = self.additive_keys(round);
+        for (i, elem) in state.iter_mut().enumerate() {
+            S::add_const_assign(elem, &keys[i], compiler);
+            S::apply_sbox(elem, compiler);
+        }
+        self.mds_matrix_multiply(state, compiler);
+    }
+
+    /// Computes a partial round at the given `round` index on the internal permutation `state`.
+    #[inline]
+    pub fn partial_round(&self, round: usize, state: &mut State<S, COM>, compiler: &mut COM) {
+        let keys = self.additive_keys(round);
+        for (i, elem) in state.iter_mut().enumerate() {
+            S::add_const_assign(elem, &keys[i], compiler);
+        }
+        S::apply_sbox(&mut state.0[0], compiler);
+        self.mds_matrix_multiply(state, compiler);
+    }
+
+    /// Computes the full permutation without the first round.
+    #[inline]
+    fn permute_without_first_round(&self, state: &mut State<S, COM>, compiler: &mut COM) {
+        for round in 1..Self::HALF_FULL_ROUNDS {
+            self.full_round(round, state, compiler);
+        }
+        for round in Self::HALF_FULL_ROUNDS..(Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS) {
+            self.partial_round(round, state, compiler);
+        }
+        for round in
+            (Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS)..(S::FULL_ROUNDS + S::PARTIAL_ROUNDS)
+        {
+            self.full_round(round, state, compiler);
+        }
+    }
+
+    /// Computes the first round borrowing the `input` and `domain_tag` returning the [`State`]
+    /// after the first round. This method does not check that `N + 1 = S::WIDTH`.
+    #[inline]
+    fn first_round_with_domain_tag_unchecked<const N: usize>(
+        &self,
+        domain_tag: &S::Field,
+        input: [&S::Field; N],
+        compiler: &mut COM,
+    ) -> State<S, COM> {
+        let mut state = Vec::with_capacity(S::WIDTH);
+        for (i, point) in iter::once(domain_tag).chain(input).enumerate() {
+            let mut elem = S::add_const(point, &self.additive_round_keys[i], compiler);
+            S::apply_sbox(&mut elem, compiler);
+            state.push(elem);
+        }
+        let mut state = State(state.into_boxed_slice());
+        self.mds_matrix_multiply(&mut state, compiler);
+        state
```

Comment:
I believe we can call assign the initial state and call `full_round` instead? This can reduce code complexity quite a bit